### PR TITLE
feat: S3 storage backend

### LINKED ISSUE #621: Implement S3 storage backend

### PR DESCRIPTION
**Pull request by jcrossley3**
Fixes #621

### CODE REVIEW

File: modules/storage/src/service/mod.rs [hunk 1]
```diff
@@ -60,10 +62,15 @@ impl TryFrom<Vec<Id>> for StorageKey {
 
 #[derive(Clone, Debug)]
 pub struct StorageResult {
-    pub key: StorageKey,
     pub digests: Digests,
 }
 
+impl StorageResult {
+    pub fn key(&self) -> StorageKey {
+        StorageKey(self.digests.sha256.encode_hex())
```
Comment (jcrossley3) [inline modules/storage/src/service/mod.rs:70]: this is the convention I wanted to keep DRY, so no longer passing in a key, but only deriving it from the digests in a single place.

File: modules/storage/src/service/s3.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+impl S3Backend {}
+
+impl S3Backend {
+    pub fn new() -> Result<Self, S3Error> {
+        // TODO: not this
+        let bucket = Bucket::new(
+            "trustify-jcrossley",
+            Region::UsEast1,
+            Credentials::default().unwrap(),
+        )?;
```
Comment (jcrossley3) [inline modules/storage/src/service/s3.rs:24]: temporary hack until i add fields to `StorageConfig`

File: modules/storage/src/service/s3.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+        // Rename the object with a key calculated from its digests
+        self.bucket
+            .copy_object_internal(&path, &key)
+            .await
+            .map_err(StoreError::Backend)?;
+        self.bucket
+            .delete_object(&path)
+            .await
+            .map_err(StoreError::Backend)?;
```
Comment (jcrossley3) [inline modules/storage/src/service/s3.rs:69]: copying/deleting on S3 for now. we can do this on the filesystem if we think it'll be too expensive.

File: modules/storage/src/service/dispatch.rs [hunk 4]
```diff
@@ -43,6 +47,11 @@ impl StorageBackend for DispatchBackend {
                 .await
                 .map(|stream| stream.map(|stream| stream.map_err(anyhow::Error::from)))
                 .map_err(anyhow::Error::from),
+            Self::S3(backend) => backend
+                .retrieve(key)
+                .await
+                .map(|stream| stream.map(|stream| stream.map_err(anyhow::Error::from)))
+                .map_err(anyhow::Error::from),
```
Comment (jcrossley3) [inline modules/storage/src/service/dispatch.rs:54]: this is where i could use some help/advice. rust compiler doesn't like seeing the different opaque types implementing `stream`. not sure of the best course here.

**Review by jcrossley3 — COMMENTED** [review]
@ctron could use your help on the compiler error in `dispatch.rs`

File: modules/storage/src/service/dispatch.rs (hunk 4 above)
Comment (bobmcwhirter) [reply modules/storage/src/service/dispatch.rs]: See subsequent commit I added for ya.

**Comment by bobmcwhirter** [comment]
Maybe start here with TRUSTD_ env prefix?

File: modules/storage/src/service/mod.rs (hunk 1 above)
Comment (ctron) [reply modules/storage/src/service/mod.rs:70]: Makes sense!

**Review by ctron — APPROVED** [review]
Looks good! And the changes to the key make perfect sense!

Not blocking the PR, but I would prefer a better way to handle the configuration switch between FS and S3. As far as I understand it, right now I could select FS (which also is the default) and provide S3 config options. And it would silently ignore the S3 config. I think there should be a away to conflict S3 options with the FS strategy (or vice versa).

**Comment by jcrossley3** [comment]
> Maybe start here with TRUSTD_ env prefix?

Done

**Pull request merged**